server/src/main/java/io/crate/analyze/ScalarsAndRefsToTrue.java

```diff
@@ -51,69 +53,77 @@
  * </pre>
  * <p>
  * <pre>
- *     true and id is null  -&gt; true and true
+ *     true and id is null  -&gt; true and null IS NULL
  * </pre>
  */
-public final class ScalarsAndRefsToTrue extends SymbolVisitor<Void, Symbol> {
+public final class ScalarsAndRefsToTrue extends SymbolVisitor<Boolean, Symbol> {
 
     private static final ScalarsAndRefsToTrue INSTANCE = new ScalarsAndRefsToTrue();
 
-    private ScalarsAndRefsToTrue() {
-    }
+    private ScalarsAndRefsToTrue() {}
 
     public static Symbol rewrite(Symbol symbol) {
-        return symbol.accept(INSTANCE, null);
+        return symbol.accept(INSTANCE, Boolean.FALSE);
     }
 
     @Override
-    public Symbol visitFunction(Function symbol, Void context) {
+    public Symbol visitFunction(Function symbol, Boolean isNullPredicate) {
         String functionName = symbol.name();
 
+        if (functionName.equals(IsNullPredicate.NAME)) {
+            isNullPredicate = Boolean.TRUE;
+        }
         if (functionName.equals(NotPredicate.NAME)) {
-            Symbol argument = symbol.arguments().get(0);
+            Symbol argument = symbol.arguments().getFirst();
             if (argument instanceof Reference) {
-                return argument.accept(this, context);
+                return argument.accept(this, isNullPredicate);
             } else if (argument instanceof Function) {
                 if (!Operators.LOGICAL_OPERATORS.contains(((Function) argument).name())) {
-                    return argument.accept(this, context);
+                    return argument.accept(this, isNullPredicate);
                 }
             }
         }
 
         List<Symbol> newArgs = new ArrayList<>(symbol.arguments().size());
         boolean allLiterals = true;
-        boolean isNull = false;
+        boolean hasNullArg = false;
         for (Symbol arg : symbol.arguments()) {
-            Symbol processedArg = arg.accept(this, context);
+            Symbol processedArg = arg.accept(this, isNullPredicate);
             newArgs.add(processedArg);
             if (!processedArg.symbolType().isValueSymbol()) {
                 allLiterals = false;
             }
             if (processedArg.valueType().id() == DataTypes.UNDEFINED.id()) {
-                isNull = true;
+                hasNullArg = true;
             }
         }
-        if (allLiterals && !Operators.LOGICAL_OPERATORS.contains(functionName)) {
-            return isNull ? Literal.NULL : Literal.BOOLEAN_TRUE;
+
+        if (functionName.equals(IsNullPredicate.NAME)) {
+            isNullPredicate = Boolean.FALSE;
+        }
```

Comment:
This is not needed (no usages of `isNullPredicate` afterwards) and would even be wrong as I think the intention would be to reset the value to it's original one.
```suggestion
```

Reply:
Yes, I was playing around with lots of things and this is a leftover, thx!